consensus/core/src/threshold_clock.rs

```diff
@@ -24,33 +26,59 @@ impl ThresholdClock {
             aggregator: StakeAggregator::new(),
             round,
             quorum_ts: Instant::now(),
+            proposal_leaders_ts: None,
             context,
         }
     }
 
     /// Add the block references that have been successfully processed and advance the round accordingly. If the round
     /// has indeed advanced then the new round is returned, otherwise None is returned.
-    pub(crate) fn add_blocks(&mut self, blocks: Vec<BlockRef>) -> Option<Round> {
+    pub(crate) fn add_blocks(
+        &mut self,
+        blocks: Vec<BlockRef>,
+        proposal_leaders_exist: bool,
+    ) -> Option<Round> {
         let previous_round = self.round;
         for block_ref in blocks {
-            self.add_block(block_ref);
+            self.add_block(block_ref, proposal_leaders_exist);
         }
         (self.round > previous_round).then_some(self.round)
     }
 
-    pub(crate) fn add_block(&mut self, block: BlockRef) {
+    pub(crate) fn get_round(&self) -> Round {
+        self.round
+    }
+
+    pub(crate) fn get_quorum_ts(&self) -> Instant {
+        self.quorum_ts
+    }
+
+    pub(crate) fn get_proposal_leaders_ts(&self) -> Option<Instant> {
+        self.proposal_leaders_ts
+    }
+
+    fn add_block(&mut self, block: BlockRef, proposal_leaders_exist: bool) {
         match block.round.cmp(&self.round) {
             // Blocks with round less then what we currently build are irrelevant here
             Ordering::Less => {}
             // If we processed block for round r, we also have stored 2f+1 blocks from r-1
             Ordering::Greater => {
                 self.aggregator.clear();
                 self.aggregator.add(block.author, &self.context.committee);
+                if proposal_leaders_exist {
```

Comment:
I feel if we want to monitor the time between round N having quorum and round N leader exists, doing it in core is more natural, and it is almost equivalent to how `block_proposal_leader_wait_ms` is computed.